dask/array/creation.py

```diff
@@ -566,7 +574,11 @@ def _diag_len(dim1, dim2, offset):
     chunks = left_chunks + right_shape
 
     graph = HighLevelGraph.from_collections(name, dsk, dependencies=[a])
-    return Array(graph, name, shape=shape, chunks=chunks, dtype=a.dtype)
+    if hasattr(a, '_meta'):
+        meta = a._meta[tuple(slice(0, 0, None) for _ in range(a.ndim))].reshape((0, ) * len(shape))
+    else:
+        meta = a[tuple(slice(0, 0, None) for _ in range(a.ndim))].reshape((0, ) * len(shape))
```

Comment:
Agreed that it would be nice to avoid having to do this all the time :)

In this particular case we could probably normalize things within the Array constructor, but I suspect that we'll need some other sort of `normalize_meta(x, shape)` function.

Reply:
Constructing `_meta` certainly can't be done in the Array constructor. Maybe normalizing dimensions could, I'm not sure yet if we have the information necessary for normalizing in all cases.

In fact, I'm not sure yet what are all the cases that can be moved to a new function or to the Array constructor, that's why I didn't want to generalize this yet, but I will do that soon.

Reply:
I think that this is something that we should figure out before merging this.

Reply:
For here we have `meta_from_array`, as I've mentioned in one of the comments from before. I just didn't go through all of the instances that need to be replaced, but will do that.